sklearn/_loss/loss.py

```diff
@@ -619,6 +620,88 @@ def fit_intercept_only(self, y_true, sample_weight=None):
+    """Huber loss, for regression.
+
+    Domain:
+    y_true and y_pred all real numbers
+    quantile in (0, 1)
+
+    Link:
+    y_pred = raw_prediction
+
+    For a given sample x_i, the Huber loss is defined as::
+
+        loss(x_i) = 1/2 * abserr**2            if abserr <= delta
+                    delta * (abserr - delta/2) if abserr > delta
+
+        abserr = |y_true_i - raw_prediction_i|
+        delta = quantile(abserr, self.quantile)
+
+    Note: HuberLoss(quantile=1) equals HalfSquaredError and HuberLoss(quantile=0)
+    equals delta * (AbsoluteError() - delta/2).
+
+    Additional Attributes
+    ---------------------
+    quantile : float
+        The quantile level which defines the breaking point `delta` to distinguish
+        between absolute error and squared error. Must be in range (0, 1).
+
+     Reference
+    ---------
+    .. [1] Friedman, J.H. (2001). :doi:`Greedy function approximation: A gradient
+      boosting machine <10.1214/aos/1013203451>`.
+      Annals of Statistics, 29, 1189-1232.
+    """
+
+    differentiable = False
+    need_update_leaves_values = True
+
+    def __init__(self, sample_weight=None, quantile=0.9, delta=0.5):
+        check_scalar(
+            quantile,
+            "quantile",
+            target_type=numbers.Real,
+            min_val=0,
+            max_val=1,
+            include_boundaries="neither",
+        )
+        self.quantile = quantile  # This is better stored outside of Cython.
+        super().__init__(
+            closs=CyHuberLoss(delta=float(delta)),
+            link=IdentityLink(),
+        )
+        self.approx_hessian = True
+        self.constant_hessian = False
+
+    def set_delta(self, y_true, raw_prediction, sample_weight=None):
+        """Calculate and set self.closs.delta based on self.quantile."""
+        abserr = np.abs(y_true - raw_prediction)
+        if sample_weight is None:
+            delta = np.percentile(abserr, 100 * self.quantile, axis=0)
+        else:
+            delta = _weighted_percentile(abserr, sample_weight, 100 * self.quantile)
+        self.closs.delta = float(delta)
+
+    def fit_intercept_only(self, y_true, sample_weight=None):
+        """Compute raw_prediction of an intercept-only model.
+
+        This is the weighted median of the target, i.e. over the samples
+        axis=0.
+        """
+        # See formula before algo 4 in Friedman (2001), but we apply it to y_true,
+        # not to the residual y_true - raw_prediction. An estimator like
+        # HistGradientBoostingRegressor might then call it on the residual, e.g.
+        # fit_intercept_only(y_true - raw_prediction).
+        if sample_weight is None:
+            median = np.percentile(y_true, 50, axis=0)
+        else:
+            median = _weighted_percentile(y_true, sample_weight, 50)
+        diff = y_true - median
+        term = np.sign(diff) * np.minimum(self.closs.delta, np.abs(diff))
+        return median + np.average(term, weights=sample_weight)
```

Comment:
Should this average be a weighted? From looking at the current GB losses, is uses a unweighted mean:

https://github.com/scikit-learn/scikit-learn/blob/1834cd6b76f63156b786e3e63cc48604532505c1/sklearn/ensemble/_gb_losses.py#L490-L492

Reply:
I think that is should be weighted and that the current GB might have an error.
Even if that's false, we can later decide to call with sample_weight=None.

Reply:
My argument is mainly based on looking at "Greedy Function Approximation" and mentally adding sample_weights everywhere, i.e. the same as for the other losses.

Reply:
And an even much better argument is the test `test_loss_intercept_only`.